proposals/handler.go

```diff
@@ -0,0 +1,380 @@
+		h.logger.WithContext(ctx).With().Error("failed to process ballot gossip", log.Err(err))
+		return pubsub.ValidationIgnore
+	}
+	return pubsub.ValidationAccept
+}
+
+// HandleBallotData handles Ballot data from gossip and sync.
+func (h *Handler) HandleBallotData(ctx context.Context, data []byte) error {
+	logger := h.logger.WithContext(ctx)
+	logger.Info("processing ballot")
+
+	var b types.Ballot
+	if err := types.BytesToInterface(data, &b); err != nil {
+		logger.With().Error("malformed ballot", log.Err(err))
+		return errMalformedData
+	}
+
+	// set the ballot and smesher ID when received
+	if err := b.Initialize(); err != nil {
+		logger.With().Error("failed to initialize ballot", log.Err(err))
+		return errInitialize
+	}
+
+	logger = logger.WithFields(b.ID(), b.LayerIndex)
+	return h.processBallot(ctx, &b, logger)
+}
+
+func (h *Handler) processProposal(ctx context.Context, data []byte) error {
+	logger := h.logger.WithContext(ctx)
+	logger.Info("processing proposal")
+
+	var p types.Proposal
+	if err := types.BytesToInterface(data, &p); err != nil {
+		logger.With().Error("malformed proposal", log.Err(err))
+		return errMalformedData
+	}
+
+	// set the proposal ID when received
+	if err := p.Initialize(); err != nil {
+		logger.With().Warning("failed to initialize proposal", log.Err(err))
+		return errInitialize
+	}
+
+	logger = logger.WithFields(p.ID(), p.Ballot.ID(), p.LayerIndex)
+
+	if h.mesh.HasProposal(p.ID()) {
+		logger.Info("known proposal")
+		return nil
+	}
+	logger.With().Info("new proposal", p.Fields()...)
+
+	if err := h.processBallot(ctx, &p.Ballot, logger); err != nil {
+		logger.With().Warning("failed to process ballot", log.Err(err))
+		return err
+	}
+
+	if err := h.checkTransactions(ctx, &p); err != nil {
+		logger.With().Warning("failed to fetch proposal TXs", log.Err(err))
+		return err
+	}
+
+	h.logger.WithContext(ctx).With().Debug("proposal is syntactically valid")
+	reportProposalMetrics(&p)
+
+	// TODO: add Proposal/Ballot/TXs to mesh
+
+	if err := h.mesh.AddProposal(&p); err != nil {
+		logger.With().Error("failed to save proposal", log.Err(err))
+		return fmt.Errorf("save proposal: %w", err)
+	}
+
+	return nil
+}
+
+func (h *Handler) processBallot(ctx context.Context, b *types.Ballot, logger log.Log) error {
+	if h.mesh.HasBallot(b.ID()) {
+		logger.Info("known ballot")
+		return nil
+	}
+	logger.With().Info("new ballot", b.Fields()...)
+
+	if err := h.checkBallotSyntacticValidity(ctx, b); err != nil {
+		logger.With().Error("ballot syntactically invalid", log.Err(err))
+		return fmt.Errorf("syntactic-check ballot: %w", err)
+	}
+
+	reportBallotMetrics(b)
+	return nil
+}
+
+func (h *Handler) checkBallotSyntacticValidity(ctx context.Context, b *types.Ballot) error {
+	h.logger.WithContext(ctx).With().Debug("checking proposal syntactic validity")
+
+	if err := h.checkBallotDataIntegrity(b); err != nil {
+		h.logger.WithContext(ctx).With().Error("ballot integrity check failed", log.Err(err))
```

Comment:
this and following logs should be wanings. receiving invalid data doesn't indicate any problems with protocol or node

Reply:
done